src/main/java/org/opentripplanner/apis/gtfs/PatternByServiceDatesFilter.java

```diff
@@ -0,0 +1,103 @@
+package org.opentripplanner.apis.gtfs;
+
+import java.time.LocalDate;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.opentripplanner.apis.gtfs.generated.GraphQLTypes.GraphQLServiceDateFilterInput;
+import org.opentripplanner.transit.model.network.Route;
+import org.opentripplanner.transit.model.network.TripPattern;
+import org.opentripplanner.transit.model.timetable.Trip;
+import org.opentripplanner.transit.service.TransitService;
+
+/**
+ * Encapsulates the logic to filter patterns by the service dates that they operate on. It also
+ * has a method to filter routes by checking if their patterns operate on the required days
+ * <p>
+ * Once a more complete filtering engine is in place in the core data model, this code should be
+ * there rather than a separate class in the API package.
+ */
+public class PatternByServiceDatesFilter {
+
+  private final LocalDate startInclusive;
+  private final LocalDate endExclusive;
+  private final Function<Route, Collection<TripPattern>> getPatternsForRoute;
+  private final Function<Trip, Collection<LocalDate>> getServiceDatesForTrip;
+
+  /**
+   * This method is not private to enable unit testing.
+   * <p>
+   * See the API documentation for a discussion of {@code startInclusive} and {@code endExclusive}.
+   */
+  PatternByServiceDatesFilter(
+    @Nullable LocalDate startInclusive,
+    @Nullable LocalDate endExclusive,
```

Comment:
These could be in a `LocalDatePeriod` or `LocalDateRange`.

Reply:
I implemented `LocalDateRange` in 0edc190f22bb03c9d97044a2c8147ed691195b37